cmd/state-installer/cmd.go

```diff
@@ -243,6 +243,13 @@ func execute(out output.Outputer, cfg *config.Instance, an analytics.Dispatcher,
 		}
 	}
 
+	// If we are installing a different branch, update install path
+	var switchChannel bool
+	if filepath.Base(params.path) != constants.BranchName {
+		params.path = filepath.Join(filepath.Dir(params.path), constants.BranchName)
+		switchChannel = true
+	}
```

Comment:
This feels too implicit. This is assuming that the path param is always implicitly given and it assumes that the installation path always ends in the branch name.

Both assumptions might currently be correct, but they may not always be. I'm not against doing this so long as we implement this in a way that clearly communicates the behavior. 

I suggest:

- Can we get rid of the path param so that we don't need this implicit logic in the first place? If not; why is it needed and how does that impact our use-case here?
- Implement a `installation.BranchForInstallPath()` function that returns which branch the given path is for. It can use the `filepath.Base()` approach; point is that it's centralized and clearly conveys the intend.

Reply:
More importantly though; I feel like this logic being in the installer at all is an issue. The `--set-channel` flag handling logic should instrument the installer in a way that provides us with the desired behavior without the installer needing to have any special knowledge over that behavior.

Reply:
> More importantly though; I feel like this logic being in the installer at all is an issue. The --set-channel flag handling logic should instrument the installer in a way that provides us with the desired behavior without the installer needing to have any special knowledge over that behavior.

This is what I initially went with but changed my approach as it would not affect previous versions of the state tool. So if you ran `state-update` with a version of the state tool that didn't have this fix it wouldn't change the installation directory. When the state tool updates itself it downloads the new installer so we can apply this fix to all of the older versions as well. If that's not an issue then we can go with that approach, however, we will have to disable the `update` integration tests for now as they are testing specific older versions of the state tool.

I like all the other suggestions in your first comment and will start looking into them 👍

Reply:
That's fine. Again this is mainly for internal use, and we wouldn't be using this with older State Tool versions once the fix lands.

The installer using the latest to perform the installation is to work around issues where the old installer is outright failing. It is not a mechanism intended to fix bugs in older State Tool versions unless it's something that's outright breaking updates.